### PR TITLE
IGNITE-11035 Updates in queries page.

### PR DESCRIPTION
**Pull request by vsisko**

### CODE REVIEW

File: modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts [hunk 1]
```diff
@@ -2058,21 +2134,76 @@ export class NotebookCtrl {
             }
         };
 
-        $window.addEventListener('beforeunload', () => {
-            this._closeOpenedQueries(this.$scope.notebook.paragraphs);
+        const offTransitions = $transitions.onBefore({from: 'base.sql.notebook'}, ($transition$) => {
```
Comment (Klaster1) [inline modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: Did you mean `this.offTransitions = `?

File: modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts [hunk 2]
```diff
@@ -2058,21 +2134,76 @@ export class NotebookCtrl {
...
+            const options = $transition$.options();
+
+            if (options.custom.justIDUpdate || options.redirectedFrom)
```
Comment (Klaster1) [inline modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: What's the point of this check? Looks like a copy/paste to me.

File: modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts [hunk 3]
```diff
@@ -2130,9 +2261,13 @@ export class NotebookCtrl {
 
     async removeParagraph(paragraph: Paragraph) {
         try {
-            await this.Confirm.confirm('Are you sure you want to remove query: "' + paragraph.name + '"?');
-            this.$scope.stopRefresh(paragraph);
+            const msg = (this._hasRunningQueries([paragraph])
+                ? 'Query is being executed. Are you sure you want cancel and remove query: "'
```
Comment (Klaster1) [inline modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: "you want _to_ cancel"

File: modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts [hunk 4]
```diff
@@ -2130,9 +2261,13 @@ export class NotebookCtrl {
 
     async removeParagraph(paragraph: Paragraph) {
```
Comment (Klaster1) [inline modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: When removing instances of Paragraph, you might want to dispose of Subjects by calling [`Subject.complete`](https://github.com/ReactiveX/rxjs/blob/6.4.0/src/internal/Subject.ts#L91).

File: modules/web-console/frontend/app/modules/agent/AgentManager.service.js [hunk 5]
```diff
@@ -723,7 +739,7 @@ export default class AgentManager {
      * @param {Number} pageSize
      * @param {Boolean} [lazy] query flag.
      * @param {Boolean} [collocated] Collocated query.
-     * @returns {Promise}
+     * @returns {Promise<VisorQueryResult>}
```
[the remark this answers is not in the document]
Comment (vsisko) [reply modules/web-console/frontend/app/modules/agent/AgentManager.service.js]: VisorQueryResult is a Java class and doesn't has analog in JS.

[the remark this answers is not in the document]
Comment (Klaster1) [reply modules/web-console/frontend/app/modules/agent/AgentManager.service.js]: That's not how types in JSDoc or TS work. You are supposed to describe the data structure so tooling and fellow developers are aware what's inside. This concerns all data received from backend.

File: modules/web-console/frontend/app/components/cluster-selector/controller.js [hunk 6]
```diff
@@ -63,8 +64,13 @@ export default class {
             this.clusters$.unsubscribe();
     }
 
-    change() {
-        this.agentMgr.switchCluster(this.cluster);
+    change(item) {
+        this.agentMgr.switchCluster(this.cluster)
+            .then(() => this.cluster = item)
```
[the remark this answers is not in the document]
Comment (vsisko) [reply modules/web-console/frontend/app/components/cluster-selector/controller.js]: We set to a cluster value of change function argument. Not result of switchCluster.
Removed true result on resolve of switchCluster function

File: modules/web-console/frontend/app/modules/agent/AgentManager.service.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (vsisko) [reply modules/web-console/frontend/app/modules/agent/AgentManager.service.js]: The returned type description has been added.

File: modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts (hunk 1 above)
Comment (vsisko) [reply modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: Changed to this.offTransitions

File: modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts (hunk 2 above)
Comment (vsisko) [reply modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: Removed unused check.

File: modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts (hunk 3 above)
Comment (vsisko) [reply modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: Typo is fixed.

File: modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts (hunk 4 above)
Comment (vsisko) [reply modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: Added closing of the subjects.

File: modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts [hunk 7]
```diff
@@ -2147,10 +2149,10 @@ export class NotebookCtrl {
             }
         };
 
-        const offTransitions = $transitions.onBefore({from: 'base.sql.notebook'}, ($transition$) => {
+        this.offTransitions = $transitions.onBefore({from: 'base.sql.notebook'}, ($transition$) => {
             const options = $transition$.options();
 
-            if (options.custom.justIDUpdate || options.redirectedFrom)
+            if (options.redirectedFrom)
```
Comment (Klaster1) [inline modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: Please add comment as to why `redirectedFrom` matters.

File: modules/web-console/frontend/app/modules/agent/AgentManager.service.js [hunk 8]
```diff
@@ -692,7 +692,7 @@ export default class AgentManager {
      * @param {Number} pageSize
      * @param {Boolean} [lazy] query flag.
      * @param {Boolean} [collocated] Collocated query.
-     * @returns {Promise<VisorQueryResult>}
+     * @returns {Promise.<{{responseNodeId: String, queryId: String, columns: String[], rows: {List.<Object[]>}, hasMore: Boolean, duration: Number}}>} Query execution result.
```
Comment (Klaster1) [inline modules/web-console/frontend/app/modules/agent/AgentManager.service.js]: 1. You can reuse type definitions using @typedef or import from TS file. Ask me how to.
2. Surely there's no `List` in parsed JSON. Please use correct types.

File: modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts (hunk 7 above)
Comment (vsisko) [reply modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: Added comment.

File: modules/web-console/frontend/app/modules/agent/AgentManager.service.js (hunk 8 above)
Comment (vsisko) [reply modules/web-console/frontend/app/modules/agent/AgentManager.service.js]: Fixed jsdock.

File: modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts [hunk 9]
```diff
@@ -1233,46 +1264,109 @@ export class NotebookCtrl {
...
+         * @param {(res) => any} onQueryStarted Action to execute when query ID is received.
+         * @return {Observable<VisorQueryResult>} Observable with first query result page.
+         */
+        const _executeQuery0 = (qryType, qryArg, onQueryStarted: (res) => any = () => {}) => {
+            return from(qryType === 'scan' ? agentMgr.queryScan(qryArg) : agentMgr.querySql(qryArg)).pipe(
+                tap((res) => {
+                    onQueryStarted(res);
+                    $scope.$applyAsync();
+                }),
+                expand((res) => {
```
Comment (nva) [inline modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: I think that an expand, filter  is not really needed and can be replaced with something like this: 
```
exhaustMap((res) => {
   if (!_.isNil(res.rows))
      return from(res);

   return merge(fetchFirstPageTask, pingQueryTask);
}) 
```

@Klaster1 what do you think about this?

File: modules/core/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryPingTask.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+            if (holder.getErr() != null)
+                return new VisorEither<>(new VisorExceptionWrapper(holder.getErr()));
+
+            holder.setAccessed(true);
+
+            if (debug)
+                log(ignite.log(), "Ping of query finished: " + qryId, getClass(), start);
+
+            return new VisorEither<>(new VisorQueryPingTaskResult(true));
```
Comment (nva) [inline modules/core/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryPingTask.java]: The inProgress flag is always true, I think it should be removed

Comment (vsisko) [reply modules/core/src/main/java/org/apache/ignite/internal/visor/query/VisorQueryPingTask.java]: Removed unused task result property.

File: modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts (hunk 9 above)
Comment (Klaster1) [reply modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: Yeah, I'm not sure recursion is even require here. The `exhaust` block basically polls first page results and pings to keep the polling alive. If I understand the query API correctly, `res.queryId` - the only value received from previous recursion call - does not change ever.

File: modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts [hunk 11]
```diff
@@ -1233,46 +1264,109 @@ export class NotebookCtrl {
...
-                take(1),
-                map((res) => _fetchQueryResult(paragraph, false, res, qryArg)),
-                catchError((err) => {
-                    if (paragraph.subscription) {
-                        paragraph.subscription.unsubscribe();
-                        paragraph.setError(err);
-
-                        _showLoading(paragraph, false);
-                        delete paragraph.subscription;
+                first()
```
Comment (Klaster1) [inline modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: @vsisko  given that `first` and `take(1)` [are slightly different](https://stackoverflow.com/q/42345969/333777), do you really intend to throw if no results arrive upon observable completion? If user cancels the query by leaving the screen, wouldn't it throw?

Comment (Klaster1) [reply modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: I'm just curious here.

Comment (vsisko) [reply modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: On cancel of the query no error is thrown.

File: modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts (hunk 9 above)
Comment (vsisko) [reply modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: Recursion is removed by changing of exexpand to exhaustMap.

**Pull request closed without merging**